src/net/sourceforge/kolmafia/persistence/ConcoctionDatabase.java

```diff
@@ -1252,6 +1253,58 @@ public static final void refreshConcoctions(boolean force) {
     ConcoctionDatabase.refreshConcoctionsNow();
   }
 
+  /**
+   * Get first index of usableList that has sortOrder non-NONE, assuming the list is sorted by
+   * sortOrder.
+   */
+  private static OptionalInt getFirstIndexNotNone() {
+    int low = 0;
+    int high = ConcoctionDatabase.usableList.size();
+    while (low < high) {
+      int mid = (low + high) / 2;
+      if (ConcoctionDatabase.usableList.get(mid).sortOrder == Concoction.Priority.NONE) {
+        low = mid + 1;
+      } else {
+        high = mid;
+      }
+    }
```

Comment:
that searches for an element, instead of for an element matching some predictate. The comparator says how the list is sorted, instead of what to search for.

Reply:
I will factor this out into a utility method that describes what it is. Not copied from anywhere - just wrote this from my head, but I'll cross-check against whatever implementation is on StackOverflow.

Reply:
As an aside, implementing a binary search is something that is typically done in an introductory programming course, after which people are expected to use someone else's implementation.  Ideally if a programmer needs a search they have the skills and resources to choose an algorithm and implementation.

I endorse Collections.binarySearch(List, Comparator) as well.

Reply:
As Chris notes above, that method is not applicable here. If you have an idea for an existing implementation already accessible to mafia I'm all ears, I just haven't been able to find one after some time spent searching. That said, the solution here is probably to restructure the data altogether, which I'll think about doing as discussed in another comment thread.

Reply:
@jaadams5 I feel like that was both rude and also entirely non useful. Truly the killer combo of bad online discourse

Reply:
> @jaadams5 I feel like that was both rude and also entirely non useful. Truly the killer combo of bad online discourse

OK.  That was not my intention and so I will apologize.  From a software engineering standpoint implementing a known standard algorithm instead of using an implementation from a library is rarely a good practice.  (We can debate that elsewhere if you really think it is debatable).  I was trying to point out an in general better way without requiring it in this specific case.  I could have done that better.